lib/dirac_coarse.cpp

```diff
@@ -190,21 +193,79 @@ namespace quda {
   {
     createY(gpu_setup, mapped);
 
-    if (gpu_setup) dirac->createCoarseOp(*Y_d,*X_d,*transfer,kappa,mass,Mu(),MuFactor());
-    else dirac->createCoarseOp(*Y_h,*X_h,*transfer,kappa,mass,Mu(),MuFactor());
+    if (!gpu_setup) {
 
-    // save the intermediate tunecache after the UV and VUV tune
-    saveTuneCache();
+      dirac->createCoarseOp(*Y_h, *X_h, *transfer, kappa, mass, Mu(), MuFactor());
+      // save the intermediate tunecache after the UV and VUV tune
+      saveTuneCache();
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+
+      createYhat(gpu_setup);
+
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+
+      calculateYhat(*Yhat_h, *Xinv_h, *Y_h, *X_h, use_mma);
+
+    } else {
+
+      // TODO: The following fancy copies reduce the number of gauge field copies (from and to QUDA_MILC_GAUGE_ORDER) by 2:
+      // one for X and one for Y, both to QUDA_MILC_GAUGE_ORDER.
+      // We haven't figure out the way to correctly use these fancy copies. :(
+#if 0
```

Comment:
@maddyscientist This is related to your initial comment on reducing the number of gauge field copies. The following code creates a `QUDA_MILC_GAUGE_ORDER` copy of `X` and `Y` before feeding them to `calculateY`, and keeps them for `calculateYhat` later. Somehow it doesn't work - it produces a slightly different `Y_d` at the end, compared to the copy-when-needed version.

The following works, at the price of an additional `exchangeGhost` (there is already one in `calculateYhat` for `Y`).
```
        dirac->createCoarseOp(*Y_order, *X_order, *transfer, kappa, mass, Mu(), MuFactor());

        // save the intermediate tunecache after the UV and VUV tune
        saveTuneCache();

        Y_d->copy(*Y_order);
        X_d->copy(*X_order);

        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");

        createYhat(gpu_setup);

        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");

        calculateYhat(*Yhat_d, *Xinv_d, *Y_order, *X_order, use_mma);

        Y_d->exchangeGhost(QUDA_LINK_BIDIRECTIONAL);

```

Some help might be needed on figure out what is happening between these copies.

Reply:
@weinbe2